subsys/disk/disk_access_stm32_sdmmc.c

```diff
@@ -439,3 +445,42 @@ DEVICE_DT_INST_DEFINE(0, disk_stm32_sdmmc_init, device_pm_control_nop,
 		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
 		    NULL);
 #endif
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay)
+
+static const struct soc_gpio_pinctrl sdmmc_pins_2[] =
+	ST_STM32_DT_INST_PINCTRL(1, 0);
+
+static struct stm32_sdmmc_priv stm32_sdmmc_priv_2 = {
+	.hsd = {
+		.Instance = (SDMMC_TypeDef *)DT_INST_REG_ADDR(1),
+	},
+#if DT_INST_NODE_HAS_PROP(1, cd_gpios)
+	.cd = {
+		.name = DT_INST_GPIO_LABEL(1, cd_gpios),
+		.pin = DT_INST_GPIO_PIN(1, cd_gpios),
+		.flags = DT_INST_GPIO_FLAGS(1, cd_gpios),
+	},
+#endif
+#if DT_INST_NODE_HAS_PROP(1, pwr_gpios)
+	.pe = {
+		.name = DT_INST_GPIO_LABEL(1, pwr_gpios),
+		.pin = DT_INST_GPIO_PIN(1, pwr_gpios),
+		.flags = DT_INST_GPIO_FLAGS(1, pwr_gpios),
+	},
+#endif
+	.pclken = {
+		.bus = DT_INST_CLOCKS_CELL(1, bus),
+		.enr = DT_INST_CLOCKS_CELL(1, bits),
+	},
+	.pinctrl = {
+		.list = sdmmc_pins_2,
+		.len = ARRAY_SIZE(sdmmc_pins_2)
+	}
+};
+
+DEVICE_DT_INST_DEFINE(1, disk_stm32_sdmmc_init, device_pm_control_nop,
+		    &stm32_sdmmc_priv_2, NULL, APPLICATION,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    NULL);
+#endif
```

Comment:
Please use an instantiation macro, similarly to what is done for other multi instance drivers. 
For instance:
https://github.com/zephyrproject-rtos/zephyr/blob/5d7fc17034b29087af801577baadca7f96b64fd4/drivers/spi/spi_ll_stm32.c#L918-L950

Reply:
Tried this in the beginning, but gave up on conditional cd and pwr definitions...
Something like this would work, but gives few pages of errors when compiling. Any hints where I can find something similar as reference?
```c
#define STM32_SDMMC_GPIO_INIT(id, func)					\
	.##func = {							\
		.name = DT_INST_GPIO_LABEL(id, func##_gpios),		\
		.pin = DT_INST_GPIO_PIN(id, func##_gpios),		\
		.flags = DT_INST_GPIO_FLAGS(id, func##_gpios),		\
	}

#define STM32_SDMMC_GPIO(id, func)					\
	COND_CODE_1(DT_INST_NODE_HAS_PROP(id, func##_gpios),		\
		(STM32_SDMMC_GPIO_INIT(id, func)),			\
		(NULL))							\
	}

#define STM32_SDMMC_INIT(id)						\
									\
static const struct soc_gpio_pinctrl sdmmc_pins_##id[] =		\
	ST_STM32_DT_INST_PINCTRL(id, 0);				\
									\
static struct stm32_sdmmc_priv stm32_sdmmc_priv_##id = {		\
	.hsd = {							\
		.Instance = (SDMMC_TypeDef *)DT_INST_REG_ADDR(id),	\
	},								\
	.pclken = {							\
		.bus = DT_INST_CLOCKS_CELL(id, bus),			\
		.enr = DT_INST_CLOCKS_CELL(id, bits),			\
	},								\
	.pinctrl = {							\
		.list = sdmmc_pins_##id,				\
		.len = ARRAY_SIZE(sdmmc_pins_##id)			\
	},								\
	STM32_SDMMC_GPIO(id, cd),					\
	STM32_SDMMC_GPIO(id, pwr),					\
};									\
									\
DEVICE_DT_INST_DEFINE(id, disk_stm32_sdmmc_init, device_pm_control_nop,	\
		    &stm32_sdmmc_priv_##id, NULL, APPLICATION,		\
		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
		    NULL)

DT_INST_FOREACH_STATUS_OKAY(STM32_SDMMC_INIT)
```

Reply:
I got the following working:
```
#define STM32_SDMMC_GPIO_CONFIG(id, func)				\
		.name = DT_INST_GPIO_LABEL(id, func##_gpios),		\
		.pin = DT_INST_GPIO_PIN(id, func##_gpios),		\
		.flags = DT_INST_GPIO_FLAGS(id, func##_gpios),		\

#define STM32_SDMMC_GPIO(id, func)					\
	.func = {							\
		COND_CODE_1(DT_INST_NODE_HAS_PROP(id, func##_gpios),	\
		(STM32_SDMMC_GPIO_CONFIG(id, func)),			\
		(NULL))							\
	}

#define STM32_SDMMC_INIT(id)						\
									\
static const struct soc_gpio_pinctrl sdmmc_pins_##id[] =		\
	ST_STM32_DT_INST_PINCTRL(id, 0);				\
									\
static struct stm32_sdmmc_priv stm32_sdmmc_priv_##id = {		\
	.hsd = {							\
		.Instance = (SDMMC_TypeDef *)DT_INST_REG_ADDR(id),	\
	},								\
	.pclken = {							\
		.bus = DT_INST_CLOCKS_CELL(id, bus),			\
		.enr = DT_INST_CLOCKS_CELL(id, bits),			\
	},								\
	.pinctrl = {							\
		.list = sdmmc_pins_##id,				\
		.len = ARRAY_SIZE(sdmmc_pins_##id)			\
	},								\
	STM32_SDMMC_GPIO(id, cd),					\
	STM32_SDMMC_GPIO(id, pe),					\
};									\
									\
DEVICE_DT_INST_DEFINE(id, disk_stm32_sdmmc_init, device_pm_control_nop,	\
		    &stm32_sdmmc_priv_##id, NULL, APPLICATION,		\
		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
		    NULL);						\

DT_INST_FOREACH_STATUS_OKAY(STM32_SDMMC_INIT)

```

Though, I had to change `STM32_SDMMC_GPIO(id, pwr),` to `STM32_SDMMC_GPIO(id, pe),`.
Can you check this particular point ?

Reply:
Won't this set .func first value to null instead of whole null in case pin is not defined?
I think having defines repeated twice makes sense - too much preprocessor magic for me otherwise...

Reply:
> Won't this set .func first value to null instead of whole func in case pin is not defined?

This was a quick hack, so I can't swear it's bug free, though this kind of processing is used at other places in tree.

> I think having defines repeated twice makes sense - too much preprocessor magic for me otherwise...

I can understand this feels a lot of preprocessor treatment for a little gain.
But this is the way it is done in other drivers so it makes sense to make it consistent with other drivers. One gain is the minimized maintenance vs both instances. 
Anyway I won't block for this... but I will make sure it is done afterwards.

Reply:
Lets have a loop then, I'm fine with pe too. Possibly it is used by other devices and I did not define pwr properly before?